man/man1/extract.1
.\"
.\"	SWISH++
.\"	extract.1
.\"
.\"	Copyright (C) 1998-2026  Paul J. Lucas
.\"
.\"	This program is free software; you can redistribute it and/or modify
.\"	it under the terms of the GNU General Public License as published by
.\"	the Free Software Foundation; either version 2 of the License, or
.\"	(at your option) any later version.
.\"
.\"	This program is distributed in the hope that it will be useful,
.\"	but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"	GNU General Public License for more details.
.\"
.\"	You should have received a copy of the GNU General Public License
.\"	along with this program; if not, write to the Free Software
.\"	Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.\" ---------------------------------------------------------------------------
.\" define code-start macro
.de cS
.sp
.nf
.RS 4
.ft CW
.ta .5i 1i 1.5i 2i 2.5i 3i 3.5i 4i 4.5i 5i 5.5i
..
.\" define code-end macro
.de cE
.ft 1
.RE
.fi
.if !'\\$1'0' .sp
..
.\" ---------------------------------------------------------------------------
.TH \f3extract\fP 1 "December 29, 2015" "SWISH++"
.SH NAME
extract \- SWISH++ text extractor
.SH SYNOPSIS
.B extract
.RI [ options ]
.I directory...
.I file...
.SH DESCRIPTION
.B extract
is the SWISH++ text extractor,
a utility to extract what text there is from a (mostly) binary file
(similar to the
.BR strings (1)
command)
prior to indexing.
Original files are untouched.
.P
Text is extracted from the specified files
and files in the specified directories;
text from files in subdirectories of specified directories is also extracted
by default (unless the
.BR \-r ,
.BR \-\-no-recurse ,
.BR \-f ,
or
.B \-\-filter
option or the
.B RecurseSubdirs
or
.B ExtractFilter
variable is given).
.P
Ordinarily,
text is extracted from files either only if their filename matches
one of the patterns in the set specified with either the
.B \-e
or
.B \-\-pattern
option or the
.B IncludeFile
variable (unless standard input is used; see next paragraph)
or is not among the set specified with either the
.B \-E
or
.B \-\-no-pattern
option or the
.B ExcludeFile
variable.
.P
If there is a single filename of `\f(CW-\f1',
the list of directories and files to extract
is instead taken from standard input (one per line).
In this case,
filename patterns of files to extract need not be specified explicitly:
all files, regardless of whether they match a pattern
(unless they are among the set not to extract specified with either the
.B \-E
or
.B \-\-no-pattern
option or the
.B ExcludeFile
variable), are extracted, i.e.,
.B extract
assumes you know what you're doing when specifying filenames in this manner.
.P
Ordinarily,
the text extracted from a file is written to another file
in the same directory having the same filename
but with the ``\f(CW.txt\fP'' extension
appended by default, e.g., ``\f(CWfoo.doc\fP'' becomes ``\f(CWfoo.doc.txt\fP''
after extraction.
(See also the
.B \-x
or
.B \-\-extension
option or the
.B ExtractExtension
variable.)
However,
extraction is not performed if the extracted text file exists.
.P
If either the
.B \-f
or
.B \-\-filter
option or the
.B ExtractFilter
variable is given,
then only a single file specified on the command line
is extracted to standard output.
In this case, filename patterns are not used
and the existence of an extracted text file is irrelevant.
.SS Filters
Via the
.B FilterFile
configuration file variable,
files having particular patterns can be filtered prior to extraction.
(See the examples in
.BR swish++.conf (4).)
.SS Character Mapping and Word Determination
.B extract
performs the same character mapping, character entity conversions,
and word determination heuristics used by
.BR index (1)
but also additionally:
.TP 3
1.
Considers all PostScript Level 2 operators that are not also English words
to be stop words.
Such words in a file usually indicate an encapsulated PostScript (EPS) file
and such should not be indexed.
.TP
2.
Looks specifically for encapsulated PostScript (EPS) data between
everything between one of
\f(CW%%BeginSetup\fP,
\f(CW%%BoundingBox\fP,
\f(CW%%Creator\fP,
\f(CW%%EndComments\fP,
or
\f(CW%%Title\fP
and
\f(CW%%Trailer\fP
and discards it.
.TP
3.
Discards strings of ASCII hex data \f(CWWord_Hex_Min_Size\fP
characters or longer, e.g., ``\f(CW7F454C46\fP.''
(Default is 5.)
.SS Motivation
.B extract
was developed to be able to index non-text files in proprietary formats
such as Microsoft Office documents.
There are a couple of reasons why the functionality of
.B extract
isn't simply built into
.BR index (1):
.TP 3
1.
Users who do not need to index such documents
shouldn't have to pay the performance penalty for doing the extra checks for
PostScript and hex data.
.TP
2.
While
.BR index (1)
can uncompress files on the fly using filters also,
uncompressing them every time indexing is performed is excessive.
Text extraction,
on the other hand,
is done only once per file;
if the file is updated,
the text-extracted version should be deleted and recreated.
.SH OPTIONS
Options begin with either a `\f(CW-\f1' for short options
or a ``\f(CW--\f1'' for long options.
Either a `\f(CW-\f1' or ``\f(CW--\f1'' by itself explicitly ends the options;
however, the difference is that `\f(CW-\f1' is returned as the first non-option
whereas ``\f(CW--\f1'' is skipped entirely.
Long option names may be abbreviated
so long as the abbreviation is unambiguous.
.P
For a short option that takes an argument,
the argument is either taken to be the remaining characters of the same option,
if any, or, if not, is taken from the next option unless said option begins
with a `\f(CW-\f1'.
.P
Short options that take no arguments can be grouped
(but the last option in the group can take an argument), e.g.,
\f(CW-lrv4\fP
is equivalent to
\f(CW-l -r -v4\fP.
.P
For a long option that takes an argument,
the argument is either taken to be the characters after a `\f(CW=\fP', if any,
or, if not, is taken from the next option unless said option begins with
a `\f(CW-\fP'.
.TP 8
.BR \-? " | " \-\-help
Prints the usage (``help'') message and exits.
.TP
.BI \-c " f" "\f1 | \fP" "" \-\-config-file \f1=\fPf
The name of the configuration file,
.IR f ,
to use.
(Default is \f(CWswish++.conf\f1 in the current directory.)
A configuration file is not required:
if none is specified and the default does not exist, none is used;
however, if one is specified and it does not exist, then this is an error.
.TP
.BI \-e " p" \f1[,\fP p... \f1]\fP "\f1 | \fP" \-\-pattern \f1=\fPp \f1[,\fP p... \f1]\fP
A filename pattern
(or set of patterns separated by commas),
.IR p ,
of files to extract text from.
Case is significant.
Multiple
.B \-e
or
.B \-\-pattern
options may be specified.
.TP
.BI \-E " p" \f1[,\fP p... \f1]\fP "\f1 | \fP" \-\-no-pattern \f1=\fPp \f1[,\fP p... \f1]\fP
A filename pattern or patterns,
.IR p ,
of files
.I not
to extract text from.
Case is significant.
Multiple
.B \-E
or
.B \-\-no-pattern
options may be specified.
.TP
.BR \-f " | " \-\-filter
Extract a single file to standard output and exit.
.TP
.BR \-l " | " \-\-follow-links
Follows symbolic links during extraction.
The default is not to follow them.
(This option is not available under Microsoft Windows
since it doesn't support symbolic links.)
.TP
.BR \-r " | " \-\-no-recurse
Do not recursively extract the files in subdirectories,
that is: when a directory is encountered,
all the files in that directory are extracted
(modulo the filename patterns specified via the
.BR \-e ,
.BR \-\-pattern ,
.BR \-E ,
or
.B \-\-no-pattern
options or the
.B IncludeFile
or
.B ExcludeFile
variables) but subdirectories encountered are ignored
and therefore the files contained in them are not extracted.
(This option is most useful when specifying the directories and files to extract
via standard input.)
The default is to extract the files in subdirectories recursively.
.TP
.BI \-s " f" "\f1 | \fP" "" \-\-stop-file \f1=\fPf
The name of a file,
.IR f ,
containing the set stop-words to use instead of the built-in set.
Whitespace, including blank lines, and characters starting with \f(CW#\f1
and continuing to the end of the line (comments) are ignored.
.TP
.BR \-S " | " \-\-dump-stop
Dumps the built-in set of stop-words to standard output and exits.
.TP
.BI \-n " n" "\f1 | \fP" "" \-\-verbosity \f1=\fPn
The verbosity level,
.IR v ,
for printing additional information to standard output during extraction.
The verbosity levels, 0-4, are:
.RS
.TP 3
0
No output is generated (except for errors).
.TP
1
Only run statistics (elapsed time, number of files, word count) are printed.
.TP
2
Directories are printed as extraction progresses.
.TP
3
Directories and files are printed with a word-count for each file.
.TP
4
Same as 3 but also prints all files that are not extracted and why.
.RE
.TP 8
.BR \-V " | " \-\-version
Prints the version number of
.BR SWISH++
to standard output and exits.
.TP
.BI \-x " e" "\f1 | \fP" "" \-\-extension \f1=\fPe
The extension to append to filenames during extraction.
(It can be specified with or without the dot;
default is \f(CWtxt\f1.)
.SH CONFIGURATION FILE
The following variables can be set in a configuration file.
Variables and command-line options can be mixed.
.P
.RS 5
.PD 0
.TP 18
.B ExcludeFile
Same as
.B \-E
or
.B \-\-no-pattern
.TP
.B ExtractExtension
Same as
.B \-x
or
.B \-\-extension
.TP
.B ExtractFilter
Same as
.B \-f
or
.B \-\-filter
.TP
.B FilterAttachment
(See FILTERS in
.BR swish++.conf (4).)
.TP
.B FilterFile
(See FILTERS in
.BR swish++.conf (4).)
.TP
.B FollowLinks
Same as
.B \-l
or
.B \-\-follow-links
.TP
.B IncludeFile
Same as
.B \-e
or
.B \-\-pattern
.TP
.B RecurseSubdirs
Same as
.B \-r
or
.B \-\-no-recurse
.TP
.B StopWordFile
Same as
.B \-s
or
.B \-\-stop-file
.TP
.B Verbosity
Same as
.B \-v
or
.B \-\-verbosity
.PD
.RE
.SH EXAMPLES
.SS Extraction
To extract text from all Microsoft Office files on a web server:
.cS
cd /home/www/htdocs
extract -v3 -e '*.doc' -e '*.ppt' -e '*.xls' .
.cE
.SS Filters
(See the examples in
.BR swish++.conf (4).)
.SH EXIT STATUS
.PD 0
.IP 0
Success.
.IP 1
Error in configuration file.
.IP 2
Error in command-line options.
.IP 20
File to extract does not exist.
.IP 30
Unable to read stop-word file.
.PD
.SH CAVEATS
.TP 3
1.
Text extraction is not perfect, nor can be.
.TP
2.
As with
.BR index (1),
the word-determination heuristics employed are heavily geared for English.
Using SWISH++ as-is to extract files in non-English languages
is not recommended.
.SH FILES
.PD 0
.TP 18
\f(CWswish++.conf\f1
default configuration file name
.PD
.SH SEE ALSO
.BR index (1),
.BR search (1),
.BR strings (1),
.BR swish++.conf (4),
.BR glob (7)
.P
Adobe Systems Incorporated.
.I PostScript Language Reference Manual, 2nd ed.
Addison-Wesley, Reading, MA.
pp. 346-359.
.P
International Standards Organization.
``ISO/IEC 9945-2: Information Technology
-- Portable Operating System Interface (POSIX)
-- Part 2: Shell and Utilities,''
1993.
.SH AUTHOR
Paul J. Lucas
.RI < paul@lucasmail.org >


NEWS
SWISH++ NEWS -- history of user-visable changes.

Copyright (C) 1996-2026 Paul J. Lucas.
See the end of the file for license conditions.

* Changes in SWISH++ 7.0.1

** Multi-file filters
A bug when there is more than one file filter in a row has been fixed.

** Change directory help
The `--help` message for `index` now includes the missing -d/--chdir option.



* Changes in SWISH++ 7.0

** Now using Autoconf & Automake.
The autoconf & automake are now being used for SWISH++ configuration
and compilation.

** Added a test suite.
Doing a "make check" will now run a suite of tests to ensure SWISH++ is
producing correct results.

** Added HTML5 elements to HTML indexer.
SWISH++ now does a better job of indexing HTML5 markup.

** Improved RTF indexer.
SWISH++ now extracts the \title from RTF files and associates words in
Information Groups as meta names.

** Can now change directories just prior to indexing.
The index command now accepts a new -d command-line option or a new
ChangeDirectory configuration variable to change directories just prior to
indexing.

** Now using C++11.
Now using C++11 as the minimum required C++ version.

** Added Python documentation.
Wherever the documentaion gives Perl 5 examples now also gives Python examples.

* Changes in SWISH++ 6.1.4

** Fixed indexing of ID3 tags.
Indexing of some ID3 ags was broken; now fixed.

* Changes in SWISH++ 6.1.3

** Fixed the search -d option.
It simply didn't work; it now does.

* Changes in SWISH++ 6.1.2

** Fixed indexing of LaTeX files ending in '%'.
The core dump that resulting from indexing a LaTeX file ending in '%' (with no
newline) has been fixed.

* Changes in SWISH++ 6.1

** Added launchd(8) cooperation.
On Mac OS X systems, the search daemon can now be controlled by launchd(8).

* Changes in SWISH++ 6.0

** Added "near" searches.
Added the ability to search using "near" in queries.

* Changes in SWISH++ 5.15.4

** Fixed extraction from standard input.
The extract command has been fixed to work from standard input.

* Changes in SWISH++ 5.15.2

** Fixed search results.
Search results didn't include the last one, i.e., if there are supposed to be N
results, only N-1 were returned.

* Changes in SWISH++ 5.15

** Index file size reduction.
Numbers stored in generated index files are now more highly compressed
resulting in an overall average savings of approximately 6% in index size.

* Changes in SWISH++ 5.14.2

** Fixed escaping of shell meta-characters.
Passing filenames containing shell meta-characters to the shell for filtering
is now fixed.

* Changes in SWISH++ 5.14

** Added chkconfig support.
The searchd script now supports chkconfig.

* Changes in SWISH++ 5.13.4

** Fixed the search --max-results option.
The search command --max-results now actually includes the "max" part.

* Changes SWISH++ 5.13

** Added ID3 indexing.
Added an indexing module for ID3 tags (typically found inside MP3 files).
ID3v1.x and ID3v2.x through 2.4.0 are supported (with the exception of
encrypted frames).

** Added UTF-16 support.
Added support for decoding UTF-16 text (both big- and little-endian).

* Changes in SWISH++ 5.12.1

** Fixed negative ranks.
File ranks are no longer ever negative.

* Changes in SWISH++ 5.12

** Run-time setting of WordThreshold.
The WordThreshold parameter may now be set either in a config. file or on the
command line (as opposed to being only a compiled-in constant).  However, only
the super-user may specify a value larger than the default.

* Changes in SWISH++ 5.11.1

** Fixed handling of invalid UTF-8.
Invalid UTF-8 byte sequences are now handled better.

* Changes in SWISH++ 5.11

** Improved rank results.
The ranking result numbers (1-100) have been significantly improved: they are
much less striated now.

* Changes in SWISH++ 5.10

** Improved XHTML indexing.
Added some XHTML 2.0 elements.

* Changes in SWISH++ 5.9.6

** Improved LaTeX indexing.
Unknown LaTeX commands are now ignored (as they should have been) rather than
indexed.

* Changes in SWISH++ 5.9.4

** Improved FreeBSD support.
The core dumps for the search command under FreeBSD have been fixed.

* Changes in SWISH++ 5.9

** Improved XML search results.
Added XML schema information in search results XML output.

* Changes in SWISH++ 5.8

** Added UTF-8 and UTF-8 support.
Added the ability to decode UTF-7 and UTF-8 in mail/news messages.

** Improved "not" in queries.
Changed the query grammar to allow "not meta" rather than just "not primary".
(What this means is that, previously "not foo = bar" wasn't legal even though
it should have been.  Instead, you had to do "not (foo = bar)".)

* Changes in SWISH++ 5.7.1

** Fixed zero rank.
Ranks were sometimes printed as zero; this has been fixed.

* Changes in SWISH++ 5.7

** Added LaTex support.
LaTeX files can now be indexed directly.

** Document title whitespace improvement.
Document titles now have multiple whitespace characters squeezed into single
whitespace characters.

** Added TMPDIR support.
The index command will now use the value of the environment variable TMPDIR if
it's set as the default temporary directory.  However, the value is still
superseded by one of -T, --temp-dir, or TempDirectory if given.

* Changes in SWISH++ 5.6

** Added direct RTF support.
The text/enriched attachment indexer that was part of the Mail module was split
off into its own RTF (Rich Text Format) module so stand-alone RTF files can now
be indexed.

** Reseting TCP connections.
For the search command running as a daemon, added code to reset the TCP
connection for bad requests.  (The reason for doing this is so we don't
potentially have a socket lingering in TIME-WAIT from a client that was too
dumb to give us a valid request in the first place.  This helps alleviate
denial-of-service attacks (if that's what's going on).)

* Changes in SWISH++ 5.5.3

** Added Procmail script.
A sample Procmail recipe has been added that can be used to split incoming mail
messages into individual files for indexing.


src/extract.c
/*
**      SWISH++
**      src/extract.c
**
**      Copyright (C) 1998  Paul J. Lucas
**
**      This program is free software; you can redistribute it and/or modify
**      it under the terms of the GNU General Public License as published by
**      the Free Software Foundation; either version 2 of the License, or
**      (at your option) any later version.
**
**      This program is distributed in the hope that it will be useful,
**      but WITHOUT ANY WARRANTY; without even the implied warranty of
**      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
**      GNU General Public License for more details.
**
**      You should have received a copy of the GNU General Public License
**      along with this program; if not, write to the Free Software
**      Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

// standard
#include <cstdlib>                              /* for exit(3) */
#include <cstring>
#include <fstream>
#include <iomanip>                              /* for setfill(), setw() */
#include <iostream>
#include <string>
#include <sys/types.h>
#include <time.h>

// local
#include "platform.h"
#include "config.h"
#include "ExcludeFile.h"
#include "exit_codes.h"
#include "ExtractExtension.h"
#include "ExtractFile.h"
#include "ExtractFilter.h"
#include "FilterFile.h"
#ifndef PJL_NO_SYMBOLIC_LINKS
#include "FollowLinks.h"
#endif
#include "pjl/mmap_file.h"
#include "pjl/option_stream.h"
#include "postscript.h"
#include "RecurseSubdirs.h"
#include "StopWordFile.h"
#include "stop_words.h"
#include "util.h"
#include "Verbosity.h"
#include "version.h"
#include "word_util.h"

using namespace PJL;
using namespace std;

ExcludeFile         exclude_patterns;           // do not extract these
ExtractFile         include_patterns;           // do extract these
ExtractFilter       extract_as_filter;
ExtractExtension    extract_extension;
FilterFile          file_filters;
static bool         in_postscript;
char const*         me;                         // executable name
int                 num_examined_files;
int                 num_extracted_files;
RecurseSubdirs      recurse_subdirectories;
Verbosity           verbosity;                  // how much to print

static bool         extract_word( char *word, int len, ostream& );
static void         extract_words(
                        mmap_file::const_iterator begin,
                        mmap_file::const_iterator end,
                        ostream&
                    );
static ostream&     usage( ostream& = cerr );

#define EXTRACT
#include "do_file.c"
#include "directory.c"

//*****************************************************************************
//
// SYNOPSIS
//
        int main( int argc, char *argv[] )
//
// DESCRIPTION
//
//      Parse the command line, initialize, call other functions ... the usual
//      things that are done in main().
//
// PARAMETERS
//
//      argc    The number of arguments.
//
//      argv    A vector of the arguments; argv[argc] is null.  Aside from the
//              options below, the arguments are the names of the files and
//              directories to be extracted.
//
// SEE ALSO
//
//      Bjarne Stroustrup.  "The C++ Programming Language, 3rd ed."
//      Addison-Wesley, Reading, MA, 1997.  pp. 116-118.
//
//*****************************************************************************
{
    me = ::strrchr( argv[0], '/' );             // determine base name...
    me = me ? me + 1 : argv[0];                 // ...of executable

#ifdef  RLIMIT_CPU                              /* SVR4, 4.2+BSD */
    //
    // Max-out the amount of CPU time we can run since extraction can take a
    // while.
    //
    max_out_limit( RLIMIT_CPU );
#endif
    /////////// Process command-line options //////////////////////////////////

    static option_stream::spec const opt_spec[] = {
        "help",         0, '?',
        "config",       1, 'c',
        "pattern",      1, 'e',
        "no-pattern",   1, 'E',
        "filter",       0, 'f',
#ifndef PJL_NO_SYMBOLIC_LINKS
        "follow-links", 0, 'l',
#endif
        "no-recurse",   0, 'r',
        "stop-file",    1, 's',
        "dump-stop",    0, 'S',
        "verbose",      1, 'v',
        "version",      0, 'V',
        "extension",    1, 'x',
        0
    };

    char const*     config_file_name_arg = ConfigFile_Default;
    bool            dump_stop_words_opt = false;
    bool            extract_as_filter_opt = false;
    char const*     extract_extension_arg = 0;
#ifndef PJL_NO_SYMBOLIC_LINKS
    bool            follow_symbolic_links_opt = false;
#endif
    bool            recurse_subdirectories_opt = false;
    StopWordFile    stop_word_file_name;
    char const*     stop_word_file_name_arg = 0;
    char const*     verbosity_arg = 0;

    option_stream opt_in( argc, argv, opt_spec );
    for ( option_stream::option opt; opt_in >> opt; )
        switch ( opt ) {

            case '?': // Print help.
                cerr << usage;

            case 'c': // Specify config. file.
                config_file_name_arg = opt.arg();
                break;

            case 'e': { // Filename pattern(s) to extract.
                char *a = opt.arg();
                for ( char *pat; pat = ::strtok( a, "," ); ) {
                    include_patterns.insert( pat, 0 );
                    a = 0;
                }
                break;
            }

            case 'E': { // Filename pattern(s) not to extract.
                char *a = opt.arg();
                for ( char *pat; pat = ::strtok( a, "," ); ) {
                    exclude_patterns.insert( pat );
                    a = 0;
                }
                break;
            }

            case 'f': // Run as a filter.
                extract_as_filter_opt = true;
                break;
#ifndef PJL_NO_SYMBOLIC_LINKS
            case 'l': // Follow symbolic links during extraction.
                follow_symbolic_links_opt = true;
                break;
#endif
            case 'r': // Specify whether to extract recursively.
                recurse_subdirectories_opt = true;
                break;

            case 's': // Specify stop-word list.
                stop_word_file_name_arg = opt.arg();
                break;

            case 'S': // Dump stop-word list.
                dump_stop_words_opt = true;
                break;

            case 'v': // Specify verbosity level.
                verbosity_arg = opt.arg();
                break;

            case 'V': // Display version and exit.
                cout << "SWISH++ " << version << endl;
                ::exit( Exit_Success );

            case 'x': // Specify filename extension to append.
                extract_extension_arg = opt.arg();
                break;

            default: // Bad option.
                cerr << usage;
        }
    argc -= opt_in.shift(), argv += opt_in.shift();

    //
    // First, parse the config. file (if any); then override variables
    // specified on the command line with options.
    //
    conf_var::parse_file( config_file_name_arg );

    if ( extract_as_filter_opt )
        extract_as_filter = true;
    if ( extract_as_filter ) {
        //
        // When running as a filter, patterns aren't used.  We clear them here
        // in case some were set via an IncludeFile directive in a
        // configuration file.  That let's us get away with not having to
        // special-case the code in do_file().
        //
        exclude_patterns.clear();
        include_patterns.clear();
    }
    if ( extract_extension_arg )
        extract_extension = extract_extension_arg;
    if ( *extract_extension != '.' )            // prepend '.' if needed
        extract_extension =
            string( "." ) + (char const*)extract_extension;
#ifndef PJL_NO_SYMBOLIC_LINKS
    if ( follow_symbolic_links_opt )
        follow_symbolic_links = true;
#endif
    if ( recurse_subdirectories_opt )
        recurse_subdirectories = false;
    if ( stop_word_file_name_arg )
        stop_word_file_name = stop_word_file_name_arg;
    if ( verbosity_arg )
        verbosity = verbosity_arg;

    /////////// Deal with stop-words //////////////////////////////////////////

    stop_words = new stop_word_set( stop_word_file_name );
    if ( dump_stop_words_opt ) {
        ::copy( stop_words->begin(), stop_words->end(),
            ostream_iterator< char const* >( cout, "\n" )
        );
        ::exit( Exit_Success );
    }

    /////////// Extract specified directories and files ///////////////////////

    bool const using_stdin = *argv && (*argv)[0] == '-' && !(*argv)[1];
    if ( !( extract_as_filter || using_stdin ) &&
        include_patterns.empty() && exclude_patterns.empty()
    )
        error() << "filename patterns must be specified when not"
                   " a filter nor\nusing standard input\n" << usage;

    if ( !argc )
        cerr << usage;

    ////////// Extract text from specified files //////////////////////////////

    time_t time = ::time( 0 );                  // Go!

    if ( extract_as_filter ) {
        //
        // Do a single file.
        //
        if ( !file_exists( *argv ) ) {
            error() << *argv << " does not exist\n";
            ::exit( Exit_No_Such_File );
        }
        do_file( *argv );
    } else if ( using_stdin ) {
        //
        // Read file/directory names from standard input.
        //
        char file_name[ PATH_MAX + 1 ];
        while ( cin.getline( file_name, PATH_MAX ) ) {
            if ( !file_exists( file_name ) ) {
                if ( verbosity > 3 )
                    cout << "  " << file_name << " (skipped: does not exist)\n";
                continue;
            }
            if ( is_directory() )
                do_directory( new_strdup( file_name ) );
            else
                do_file( file_name );
        }
    } else {
        //
        // Read file/directory names from command line.
        //
        for ( ; *argv; ++argv ) {
            if ( !file_exists( *argv ) ) {
                if ( verbosity > 3 )
                    cout << "  " << *argv << " (skipped: does not exist)\n";
                continue;
            }
            if ( is_directory() )
                do_directory( *argv );
            else
                do_file( *argv );
        }
    }

    if ( verbosity ) {
        time = ::time( 0 ) - time;              // Stop!
        cout << '\n' << me << ": done:\n  "
             << setfill('0')
             << setw(2) << (time / 60) << ':'
             << setw(2) << (time % 60)
             << " (min:sec) elapsed time\n  "
             << num_examined_files << " files, "
             << num_extracted_files << " extracted\n\n";
    }

    ::exit( Exit_Success );
}

//*****************************************************************************
//
// SYNOPSIS
//
        bool extract_word( register char *word, register int len, ostream &out )
//
// DESCRIPTION
//
//      Potentially extract the given word.
//
// PARAMETERS
//
//      word    The candidate word to be extracted.
//
//      len     The length of the word since it is not null-terminated.
//
//      out     The ostream to write the word to.
//
// RETURN VALUE
//
//      Returns true only if the word was extracted.
//
//*****************************************************************************
{
    if ( len < Word_Hard_Min_Size )
        return false;

    word[ len ] = '\0';

    ////////// Look for Encapsulated PostScript code and skip it //////////////

    if ( in_postscript ) {
        if ( !::strcmp( word, "%%Trailer" ) )
            in_postscript = false;
        return false;
    }
    static postscript_comment_set const postscript_comments;
    if ( postscript_comments.contains( word ) ) {
        in_postscript = true;
        return false;
    }
    static postscript_operator_set const postscript_operators;
    if ( postscript_operators.contains( word ) )
        return false;

    ////////// Strip chars not in Word_Begin_Chars/Word_End_Chars /////////////

    for ( register int i = len - 1; i >= 0; --i ) {
        if ( is_word_end_char( word[ i ] ) )
            break;
        --len;
    }
    if ( len < Word_Hard_Min_Size )
        return false;

    word[ len ] = '\0';

    while ( *word ) {
        if ( is_word_begin_char( *word ) || *word == '%' )
            break;
        --len, ++word;
    }
    if ( len < Word_Hard_Min_Size )
        return false;

    ////////// Discard what looks like ASCII hex data /////////////////////////

    if ( len > Word_Hex_Max_Size &&
         (int)::strspn( word, "0123456789abcdefABCDEF" ) == len
    )
        return false;

    ////////// Stop-word checks ///////////////////////////////////////////////

    if ( !is_ok_word( word ) || stop_words->contains( to_lower( word ) ) )
        return false;

    out << word << '\n';
    return true;
}

//*****************************************************************************
//
// SYNOPSIS
//
        void extract_words(
            register mmap_file::const_iterator c,
            register mmap_file::const_iterator end,
            ostream &out
        )
//
// DESCRIPTION
//
//      Extract the words between the given iterators.
//
// PARAMETERS
//
//      c       The iterator marking the beginning of the text to extract.
//
//      end     The iterator marking the end of the text to extract.
//
//      out     The ostream to write the words to.
//
//*****************************************************************************
{
    char    word[ Word_Hard_Max_Size + 1 ];
    int     len;
    int     num_words = 0;
    bool    in_word = false;

    in_postscript = false;

    while ( c != end ) {
        register char const ch = *c++;

        ////////// Collect a word /////////////////////////////////////////////

        if ( is_word_char( ch ) || ch == '%' ) {
            if ( !in_word ) {                   // start a new word
                word[ 0 ] = ch;
                len = 1;
                in_word = true;
                continue;
            }
            if ( len < Word_Hard_Max_Size ) {   // continue same word
                word[ len++ ] = ch;
                continue;
            }
            in_word = false;                    // too big: skip chars
            while ( c != end && is_word_char( *c++ ) ) ;
            continue;
        }

        if ( in_word ) {
            //
            // We ran into a non-word character, so extract the word up to, but
            // not including, it.
            //
            in_word = false;
            num_words += extract_word( word, len, out );
        }
    }
    if ( in_word ) {
        //
        // We ran into 'end' while still accumulating characters into a word,
        // so just extract what we've got.
        //
        num_words += extract_word( word, len, out );
    }

    if ( verbosity > 2 )
        cout << " (" << num_words << " words)" << endl;
}

//*****************************************************************************
//
//  Miscellaneous function(s)
//
//*****************************************************************************

ostream& usage( ostream &err ) {
    err << "usage: " << me << " [options] dir ... file ...\n"
    "options: (unambiguous abbreviations may be used for long options)\n"
    "========\n"
    "-?   | --help             : Print this help message\n"
    "-c f | --config-file f    : Name of configuration file [default: " << ConfigFile_Default << "]\n"
    "-e p | --pattern p        : Filename pattern to extract [default: none]\n"
    "-E p | --no-pattern p     : Filename pattern not to extract [default: none]\n"
    "-f   | --filter           : Filter one file to standard output [default: no]\n"
#ifndef PJL_NO_SYMBOLIC_LINKS
    "-l   | --follow-links     : Follow symbolic links [default: no]\n"
#endif
    "-r   | --no-recurse       : Don't extract subdirectories [default: do]\n"
    "-s f | --stop-file f      : Stop-word file to use instead of built-in default\n"
    "-S   | --dump-stop        : Dump stop-words, exit\n"
    "-v v | --verbosity v      : Verbosity level [0-4; default: 0]\n"
    "-V   | --version          : Print version number, exit\n"
    "-x e | --extension e      : Extension to append to filename [default: txt]\n";
    ::exit( Exit_Usage );
    return err;                                 // just to make compiler happy
}
/* vim:set et sw=4 ts=4: */
